using_restricted_array.rb

```diff
@@ -82,14 +87,18 @@ def binary_search(array, length, value_to_find)
   low = 0
   high = length
 
-  while low < high
-    mid = (low + high) / 2
-    if value_to_find == array[mid]
-      return true
-    elsif value_to_find < array[mid]
-      high = mid
-    else
-      low = mid
+  if length == 1 && value_to_find == array[0]
+    return true
+  else
+    while low < high
+      mid = (low + high) / 2
+      if value_to_find == array[mid]
+        return true
+      elsif value_to_find < array[mid]
+        high = mid
+      else
+        low = mid
+      end
```

Comment:
Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, low = 0, high=1, mid=0. Value is not at index 0. So, we make low = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. To fix this, update high to be mid -1, if value_to_find is less than mid. And update low = mid+1, if value_to_find is greater than mid.
This was another consideration I was missing in the algorithm when we reviewed it in class. The slide has since been fixed.